### PR TITLE
Product kernel

### PR DESCRIPTION
**Pull request by MJHutchinson**
This implements two things:
- Producting compact manifolds
- Producting kernels

Since we haven't really set up for RFF/Euclidean kernels yet I haven't included the method for combing KL/RFF sampling - I'll do this when we have a clearer idea on that

### CODE REVIEW

File: geometric_kernels/kernels/geometric_kernels.py [hunk 1]
```diff
@@ -81,7 +80,7 @@ def _spectrum(
             return B.exp(-(lengthscale ** 2) / 2.0 * (s ** 2))
         elif nu > 0:
             power = -nu - self.space.dimension / 2.0
-            base = 2.0 * nu / lengthscale ** 2 + from_numpy(nu, s ** 2)
+            base = 2.0 * nu / lengthscale ** 2 + s ** 2  # from_numpy(nu, s ** 2)
```
Comment (stoprightthere) [inline geometric_kernels/kernels/geometric_kernels.py:83]: Why is this change needed?

File: geometric_kernels/kernels/product.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    @property
+    def space(self) -> List[Space]:
+        return [kernel.space for kernel in self.kernels]
+
+    def init_params_and_state(self) -> Tuple(List[dict], List[dict]):
+        params_and_state = [kernel.init_params_and_state() for kernel in self.kernels]
+
+        return [p[0] for p in params_and_state], [s[1] for s in params_and_state]
+
+    def K(self, params, state, X, X2=None, **kwargs) -> B.Numeric:
```
Comment (vdutor) [inline geometric_kernels/kernels/product.py:38]:
```suggestion
    def K(self, params: List[Mapping], state: List[Mapping], X, X2=None, **kwargs) -> B.Numeric:
```

Comment (vdutor) [reply geometric_kernels/kernels/product.py:38]: Adding some docstrings with expected shapes with be very helpful.

File: geometric_kernels/kernels/product.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if X2 is None:
+            X2 = X
+
+        Xs = [B.take(X, inds, axis=-1) for inds in self.dimension_indices]
+        X2s = [B.take(X2, inds, axis=-1) for inds in self.dimension_indices]
+
+        return B.stack(
+            *[
+                kernel.K(X, X2, p, s)
```
Comment (vdutor) [inline geometric_kernels/kernels/product.py:47]: I'm surprised by the order of the input arguments. Doesn't seem to match the interface: https://github.com/GPflow/GeometricKernels/blob/76254b0e36fd41fca6656f63356924db4f16b2b4/geometric_kernels/kernels/base.py#L42?

File: geometric_kernels/kernels/product.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            ],
+            axis=-1,
+        ).prod(dim=-1)
+
+    def K_diag(self, params, state, X):
+        Xs = [B.take(X, inds, axis=-1) for inds in self.dimension_indices]
+
+        return B.stack(
+            *[
+                kernel.K_diag(X, p, s)
```
Comment (vdutor) [inline geometric_kernels/kernels/product.py:58]: I'm surprised by the order of the input arguments. Doesn't seem to match the interface?
https://github.com/GPflow/GeometricKernels/blob/76254b0e36fd41fca6656f63356924db4f16b2b4/geometric_kernels/kernels/base.py#L49

File: geometric_kernels/spaces/circle.py [hunk 5]
```diff
@@ -23,14 +24,14 @@ class SinCosEigenfunctions(EigenfunctionWithAdditionTheorem):
...
+        # assert (
+        #     num_eigenfunctions % 2 == 1
+        # ), "num_eigenfunctions needs to be odd to include all eigenfunctions within a level."
+        # assert num_eigenfunctions >= 1
 
         self._num_eigenfunctions = num_eigenfunctions
-        # We know `num_eigenfunctions` is odd, therefore:
-        self._num_levels = num_eigenfunctions // 2 + 1
+        # Compute level that fully encompasses all the requested eigenfunctions in a level
+        self._num_levels = ceil(num_eigenfunctions / 2) + 1
```
Comment (vdutor) [inline geometric_kernels/spaces/circle.py:34]: Maybe write a message to std out to inform the user the actual number of eigenfunctions has been updated?

File: geometric_kernels/spaces/circle.py [hunk 6]
```diff
@@ -49,7 +50,8 @@ def __call__(self, X: B.Numeric, **parameters) -> B.Numeric:
                 values.append(const * B.cos(freq * theta))
                 values.append(const * B.sin(freq * theta))
 
-        return B.concat(*values, axis=1)  # [N, M]
+        # Possibly clip last eigenfunction
+        return B.concat(*values, axis=1)[:, : self._num_eigenfunctions]  # [N, M]
```
Comment (vdutor) [inline geometric_kernels/spaces/circle.py:54]:
```suggestion
        return B.concat(*values, axis=1)[:, :self._num_eigenfunctions]  # [N, M]
```

Comment (vdutor) [reply geometric_kernels/spaces/circle.py:54]: Can you explain when this would be necessary? I can imagine one always wants the full level?

File: geometric_kernels/spaces/product.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+from geometric_kernels.eigenfunctions import Eigenfunctions
+
+
+class ProductEigenfunctions(Eigenfunctions):
+    def __init__(
+        self,
+        dimensions: List[int],
+        eigenindicies: B.Numeric,
+        *eigenfunctions: Eigenfunctions,
```
Comment (vdutor) [inline geometric_kernels/spaces/product.py:20]: Why can this not be `List[Eigenfunctions]`? This would better match the `List[int]` of corresponding dimensions.

File: geometric_kernels/spaces/product.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            The dimensions of the spaces being producted together
+        eigenindicies : B.Numeric
+            An array mapping i'th eigenfunction of the product space to
+            the index of the eigenfunctions of the subspaces
+        eigenfunctions : Eigenfunctions
+
+        """
+        self.dimension_indices = []
+        i = 0
+        inds = B.linspace(0, sum(dimensions) - 1, sum(dimensions)).astype(int)
```
Comment (vdutor) [inline geometric_kernels/spaces/product.py:37]: Replace by `B.range`?

File: geometric_kernels/spaces/product.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+from .base import DiscreteSpectrumSpace
+
+from geometric_kernels.eigenfunctions import Eigenfunctions
+
+
+class ProductEigenfunctions(Eigenfunctions):
+    def __init__(
+        self,
+        dimensions: List[int],
+        eigenindicies: B.Numeric,
```
Comment (vdutor) [inline geometric_kernels/spaces/product.py:19]: Can this not be `List[int]`?

File: geometric_kernels/spaces/product.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+        r"""Implementation of products of discrete spectrum spaces.
+        Assumes the spaces are compact manifolds and that the eigenfunctions are the
+        eigenfunctions of the Laplace-Beltrami operator. On such a space the eigen(values/functions)
+        on the product space associated with the multiindex alpha are given by
+            lambda_alpha = \sum_i lambda_{i, alpha_i}
+            phi_alpha = \prod_i phi_{i, alpha_i}
+        where lambda_{i, j} is the j'th eigenvalue on the i'th manifold in the product
+        and phi_{i, j} is the j'th eigenfunction on the i'th manifold in the product.
+
+        The eigenfunctions of such manifolds can't in genreal be analytically ordered, and
```
Comment (vdutor) [inline geometric_kernels/spaces/product.py:80]:
```suggestion
        The eigenfunctions of such manifolds can't in general be analytically ordered, and
```

File: geometric_kernels/kernels/geometric_kernels.py (hunk 1 above)
Comment (lrozo) [reply geometric_kernels/kernels/geometric_kernels.py:83]: Same here, I did not get why this change is needed.

File: geometric_kernels/spaces/circle.py [hunk 11]
```diff
@@ -23,14 +24,14 @@ class SinCosEigenfunctions(EigenfunctionWithAdditionTheorem):
     """
 
     def __init__(self, num_eigenfunctions: int) -> None:
-        assert (
-            num_eigenfunctions % 2 == 1
-        ), "num_eigenfunctions needs to be odd to include all eigenfunctions within a level."
-        assert num_eigenfunctions >= 1
+        # assert (
```
Comment (lrozo) [inline geometric_kernels/spaces/circle.py:27]: I did not get why we do not need this assertion anymore, is it because we now have it in `_addition_theorem`

File: geometric_kernels/spaces/circle.py [hunk 12]
```diff
@@ -49,7 +50,8 @@ def __call__(self, X: B.Numeric, **parameters) -> B.Numeric:
                 values.append(const * B.cos(freq * theta))
                 values.append(const * B.sin(freq * theta))
 
-        return B.concat(*values, axis=1)  # [N, M]
+        # Possibly clip last eigenfunction
+        return B.concat(*values, axis=1)[:, : self._num_eigenfunctions]  # [N, M]
 
     def _addition_theorem(self, X: B.Numeric, X2: B.Numeric, **parameters) -> B.Numeric:
```
Comment (lrozo) [inline geometric_kernels/spaces/circle.py:56]: May we add some docstring for the arguments of the function?

File: notebooks/torus.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
+# %%
```
Comment (lrozo) [inline notebooks/torus.py:1]: Nice example! :)

**Review by lrozo — APPROVED** [review]
In general, I do not have major comments for this PR. Besides some few questions, my general comment is to improve the documentation of the code (several functions miss docstrings). I also did not go into the details of the class `ProductDiscretSpectrumSpace`, but the class is well commented. Maybe we may add some reference to point out to the math details covered here.

File: geometric_kernels/spaces/product.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+
+        # prefetch the eigenvalues of the subspaces
+        sub_space_eigenvalues = B.stack(
+            *[space.get_eigenvalues(self.num_eigen)[:, 0] for space in self.sub_spaces],
+            axis=0,
+        )
+        # prebuild array for indexing the first index of the eigenvalue array
+        first_index = B.linspace(
+            0, len(self.sub_spaces) - 1, len(self.sub_spaces)
+        ).astype(int)
```
Comment (stoprightthere) [inline geometric_kernels/spaces/product.py:113]: Use B.range?

File: geometric_kernels/spaces/product.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+            eigenvalues = sub_space_eigenvalues[first_index, curr_sub_space_idx].sum(
+                axis=1
+            )
+
+            # Compute tied smallest new eigenvalues
+            highest_eigenvalue_index = int(eigenvalues.argmin())
+            tied_eigenvalues = eigenvalues == eigenvalues[highest_eigenvalue_index]
+            tied_eigenvalues_indexes = B.linspace(
+                0, len(tied_eigenvalues) - 1, len(tied_eigenvalues)
+            ).astype(int)[tied_eigenvalues]
```
Comment (stoprightthere) [inline geometric_kernels/spaces/product.py:129]: Use B.argmin?

File: geometric_kernels/spaces/product.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+        first_index = B.linspace(
+            0, len(self.sub_spaces) - 1, len(self.sub_spaces)
+        ).astype(int)
+        i = 0
+
+        # first eigenvalue is the sum of the first eigenvalues of the individual spaces
+        curr_sub_space_idx = B.zeros(int, len(self.sub_spaces))[None, :]
+        while i < self.num_eigen:
+            # compute eigenvalues of the proposals
+            eigenvalues = sub_space_eigenvalues[first_index, curr_sub_space_idx].sum(
```
Comment (stoprightthere) [inline geometric_kernels/spaces/product.py:120]: Will this "advanced indexing" work for backends other than numpy?

**Comment by stoprightthere** [comment]
Looks fine to me, except minor fixes I proposed.

The algorithm to find the least sum combinations of eigenvalues seems fine, although it's implementation requires some head-scratching to understand that it does what it's supposed to do :) 

I suggest merging `main` into your branch to fix the CI and also check that the code works for every backend -- hopefully it won't require too much hussle to bring it on

File: geometric_kernels/kernels/geometric_kernels.py [hunk 17]
```diff
@@ -79,13 +79,13 @@ def _spectrum(
         Depends on the `lengthscale` parameters.
         """
         if nu == np.inf:
-            return B.exp(-(lengthscale**2) / 2.0 * from_numpy(lengthscale, s**2))
+            return B.exp(-(lengthscale ** 2) / 2.0 * from_numpy(lengthscale, s ** 2))
         elif nu > 0:
             power = -nu - self.space.dimension / 2.0
-            base = 2.0 * nu / lengthscale**2 + B.cast(
-                B.dtype(nu), from_numpy(nu, s**2)
+            base = 2.0 * nu / lengthscale ** 2 + B.cast(
```
Comment (stoprightthere) [inline geometric_kernels/kernels/geometric_kernels.py:85]: Here and so on, you sure you've got the most up-to-date linter? Because in the main branch, it is always `a**b`, never `a ** b` and the linter complains about the latter

**Comment by stoprightthere** [comment]
I'm still on it, it's been a long time so basically reviewing from scracth :)

But in any case, I see some unresolved comments. Could you please comment on them/fix em?

And linting errors have to be fixed as well.

File: geometric_kernels/basis.py [hunk 18]
```diff
@@ -0,0 +1,219 @@
+"""
```
Comment (vdutor) [inline geometric_kernels/basis.py:1]: This file seems like largely copy-pasted from https://github.com/GPflow/GeometricKernels/blob/main/geometric_kernels/spaces/eigenfunctions.py. Not sure why that is necessary?

File: geometric_kernels/basis.py [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+    @abc.abstractmethod
+    def __call__(self, X: B.Numeric, **parameters) -> B.Numeric:
+        """
+        :param X: points to evaluate the eigenfunctions in local coordinates, [N, D].
+            `N` is the number of points and `D` should match the dimension of the space
+            on which the eigenfunctions are defined.
+        :param parameters: any additional parameters
+        """
+
+    raise NotImplementedError
```
Comment (vdutor) [inline geometric_kernels/basis.py:75]: wrong indentation?

File: tests/kernels/test_product.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+        k_xx_single_1 = kernel_single.K(
+            params_single, state_single, grid_[..., :1], grid_[..., :1]
+        ).reshape(_GRID_SIZE, _GRID_SIZE, _GRID_SIZE, _GRID_SIZE)
+
+        k_xx_single_2 = kernel_single.K(
+            params_single, state_single, grid_[..., 1:], grid_[..., 1:]
+        ).reshape(_GRID_SIZE, _GRID_SIZE, _GRID_SIZE, _GRID_SIZE)
+
+        k_xx_product = k_xx_single_1 * k_xx_single_2
```
Comment (vdutor) [inline tests/kernels/test_product.py:47]: missing equality check?

File: notebooks/torus.py [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+%load_ext autoreload
+%autoreload 2
+# %%
+import lab as B
+import numpy as np
+from geometric_kernels.spaces import ProductDiscreteSpectrumSpace
+from geometric_kernels.spaces import Circle
+# %%
+
+circ = Circle()
```
Comment (vdutor) [inline notebooks/torus.py:11]: unused variable:
```suggestion
```

File: notebooks/product_kernel.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+from geometric_kernels.lab_extras.extras import from_numpy
+import numpy as np
+# %%
+
+from geometric_kernels.spaces import Circle, ProductDiscreteSpectrumSpace
+
+circle = Circle()
+
+# %%
+product = ProductDiscreteSpectrumSpace(circle, circle, num_eigen=11**2)
```
Comment (vdutor) [inline notebooks/product_kernel.py:16]: I would expect this to be safer:
```suggestion
product = ProductDiscreteSpectrumSpace(Circle(), Circle(), num_eigen=11**2)
```

File: geometric_kernels/kernels/product.py (hunk 3 above)
Comment (stoprightthere) [reply geometric_kernels/kernels/product.py:68]: Same

File: geometric_kernels/basis.py (hunk 18 above)
Comment (stoprightthere) [inline geometric_kernels/basis.py:1]: What's the deal with this module? Doesn't seem to be referenced anywhere and `eigenfunctions.py` is still present

File: geometric_kernels/spaces/base.py [hunk 23]
```diff
@@ -53,6 +56,23 @@ def get_eigenvalues(self, num: int) -> B.Numeric:
         raise NotImplementedError
 
 
+class DiscreteSpectrumSpaceWithAdditionTheorem(DiscreteSpectrumSpace):
```
Comment (stoprightthere) [inline geometric_kernels/spaces/base.py:59]: Why is this needed? To workaround the need specify the exact number of the eigenfunctions as opposed to the number of levels?

File: geometric_kernels/kernels/product.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+    TODO: Support combined RFF/KL expansion methods
+
+    Parameters
+    ----------
+    kernels : BaseGeometricKernel
+        kernels to compute the product for
+    dimension_indices : List[B.Numeric], optional
+        List of indices the correspond to the indices of the input that should be fed to each kernel.
+        If None, assume the each kernel takes kernel.space.dimension inputs, and that the input will
+        be a stack of this size, by default None
```
Comment (vdutor) [inline geometric_kernels/kernels/product.py:24]: Please move the docstrings to the `__init__`. I would also remove the type information as this is redundant.

File: geometric_kernels/spaces/circle.py [hunk 25]
```diff
@@ -91,6 +86,10 @@ def _addition_theorem_diag(self, X: B.Numeric, **parameters) -> B.Numeric:
         :return: Evaluate the sum of eigenfunctions on each level. Returns
             a value for each level [N, L]
         """
+        # assert (
+        #    self.complete_levels
+        # ), "num_eigenfunctions needs to be odd to include all eigenfunctions within a level."
+
```
Comment (vdutor) [inline geometric_kernels/spaces/circle.py:92]: Can we remove commented code?
```suggestion
```

File: geometric_kernels/spaces/circle.py [hunk 26]
```diff
@@ -118,6 +117,11 @@ def num_eigenfunctions_per_level(self) -> B.Numeric:
         """Number of eigenfunctions per level, [N_l]_{l=0}^{L-1}"""
         return [1 if level == 0 else 2 for level in range(self.num_levels)]
 
+    # @classmethod
+    # def from_levels(cls, num):
+    #    num_eigenfunctions = 2 * num - 1
+    #    return cls(num_eigenfunctions)
+
```
Comment (vdutor) [inline geometric_kernels/spaces/circle.py:125]: Can we remove commented code?
```suggestion
```

File: geometric_kernels/spaces/eigenfunctions.py [hunk 27]
```diff
@@ -128,10 +154,10 @@ def weighted_outerproduct(
             X2 = X
 
         sum_phi_phi_for_level = self._addition_theorem(X, X2, **parameters)  # [N, N, L]
-        weights = self._filter_weights(weights)
+        # weights = self._filter_weights(weights)
```
Comment (vdutor) [inline geometric_kernels/spaces/eigenfunctions.py:157]:
```suggestion
```

File: geometric_kernels/spaces/eigenfunctions.py [hunk 28]
```diff
@@ -153,9 +179,15 @@ def weighted_outerproduct_diag(
         :return: shape [N,]
         """
         addition_theorem_X = self._addition_theorem_diag(X, **parameters)  # [N, L]
-        weights = self._filter_weights(weights)
+        # weights = self._filter_weights(weights)
         # weights = from_numpy(addition_theorem_X, weights)
```
Comment (vdutor) [inline geometric_kernels/spaces/eigenfunctions.py:183]:
```suggestion
```

File: tests/kernels/test_product.py [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+
+        print(k_xx)
+        print("=========")
+        print(k_xx_product)
+
+        assert np.allclose(B.to_numpy(k_xx), B.to_numpy(k_xx_product))
```
Comment (vdutor) [inline tests/kernels/test_product.py:86]: I'd suggest using [numpy.testing.assert_allclose](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) as it gives better error messaging.

File: tests/kernels/test_product.py [hunk 30]
```diff
@@ -0,0 +1,86 @@
+import lab as B
+import numpy as np
+from opt_einsum import contract as einsum
+
+from geometric_kernels.kernels import MaternKarhunenLoeveKernel
+from geometric_kernels.lab_extras.extras import from_numpy
+from geometric_kernels.spaces import Circle, ProductDiscreteSpectrumSpace
+from geometric_kernels.utils.utils import chain
+
+# _TRUNCATION_LEVEL = 11 ** 2
```
Comment (vdutor) [inline tests/kernels/test_product.py:10]:
```suggestion
```

**Review by vdutor — COMMENTED** [review]
Looks good - I mainly left some stylistic comments.
